src/clj_sqs_extended/internal/receive.clj

```diff
@@ -32,44 +37,60 @@
   (let [old-in-chan (:in-chan @loop-state)]
     (swap! loop-state
            (fn [state]
              (-> state
-                 (assoc :in-chan (sqs/receive-message-channeled sqs-ext-client
-                                                                (:queue-name @loop-state)
-                                                                receive-opts))
+                 (assoc :in-chan
+                        (sqs/receive-message-channeled sqs-ext-client
+                                                       (:queue-name @loop-state)
+                                                       receive-opts))
                  (update-in [:stats :restart-count] inc)
                  (assoc-in [:stats :restarted-at] (t/now)))))
     (close! old-in-chan)))
 
-(defn- stop-receive-loop
-  [loop-state]
-  (when (:running @loop-state)
-    (log/infof "Terminating receive-loop for queue '%s'" (:queue-name @loop-state))
-    (swap! loop-state assoc :running false)
-    (swap! loop-state assoc-in [:stats :stopped-at] (t/now))
-    (close! (:in-chan @loop-state))
-    (close! (:out-chan @loop-state))
-    (:stats @loop-state)))
+(defn- error-might-be-recovered-by-restarting
+  [error]
+  (cond
+    (instance? ReflectiveOperationException error) false
+    (instance? RuntimeException error) false
+    (instance? UnknownHostException error) true
+    (instance? SocketException error) true
+    (instance? UnknownHostException error) true
+    (instance? HttpTimeoutException error) true
+    :else false))
 
 (defn- handle-message-receival-error
   [sqs-ext-client loop-state error restart-delay-seconds receive-opts]
-  (let [{:keys [this-pass-started-at] :as stats} (:stats @loop-state)]
-    (log/warn error "Received an error!"
-              (assoc stats :last-wait-duration
-                           (secs-between this-pass-started-at
-                                         (t/now))))
-    ;; WATCHOUT: Adding a restart delay so that this doesn't go into an
-    ;;           abusively tight loop if the queue listener is failing to
-    ;;           start continuously.
-    (<! (timeout (int (* restart-delay-seconds 1000))))
-    (restart-receive-loop sqs-ext-client
-                          loop-state
-                          receive-opts)))
+  (if (error-might-be-recovered-by-restarting error)
+    (let [count (get-in @loop-state [:stats :restart-count])]
+      (if (< count (:restart-limit receive-opts))
+        (do
+          (log/warnf error "Error occured while receiving message!")
+          (Thread/sleep (* 1000 restart-delay-seconds))
+          (restart-receive-loop sqs-ext-client
+                                loop-state
+                                receive-opts))
+        (log/warnf "Skipping receive-loop restart because the limit (%d) has been reached!"
+                   count)))
+    (do
+      (log/fatalf error "Unrecovereable error occured while receiving message!")
+      (stop-receive-loop loop-state))))
```

Comment:
we shouldn't raise fatal, I would replace this `do` block and just raise error back up. in case the user has their own error handling, we don't want to deprive them of the option and crash deliberately here. (again fatal halts the process)